### ISSUE TITLE
New rule: label-missing-input

### ISSUE THREAD

**Issue by marcysutton** [dequelabs/axe-core#783]
# label-missing-input

Sometimes in the interest of styling form controls like radio buttons and checkboxes, a `<label>` element is styled to appear interactive with CSS but its paired input is hidden with `display: none`. The problem with this is there is no keyboard-interactive control to focus on, so sighted keyboard users can't operate it. We should be able to look at the label and input pairing, inspect the input's CSS, and if there isn't some other rendered focusable control, we can throw a violation. 

Here's an example, it occurs on the "Time: Day/Night" toggle in the adjust options widget: https://www.blinds.com/c/wood-blinds/54
```html
<div class="vista-item" data-reactid="40">
    <input type="radio" id="time-of-day-1" name="time-of-day" value="1" style="display:none;" class="radio-label" data-reactid="41">
    <label for="time-of-day-1" class="vista-label" data-reactid="42">
        <span data-reactid="43">Night</span>
    </label>
</div>
```

**Tags:** 
wcag2a, wcag211, cat.keyboard

## Selector

`label`

### label-has-visible-input (any)

1. Label with a for/id pairing or implicit wrapping must have a rendered control to operate with the keyboard. We can check if the control has `display: none` CSS, and if so, keep looking for other inputs. Are there duplicates where one is visible, for example, for RWD? If there are none, return false.

**Comment by marcysutton** [comment]
This would probably highlight other abuses of the label element, as well! Like throwing around `<label>` without it actually being used for a form control.

**Comment by WilcoFiers** [comment]
Are there things we can say about this anti-pattern that isn't going to be true about other form elements? Because the more of that we have, the less likely this is to turn up false positives. Off the top of my head: Is this only true for toggles? For example, radio buttons with just two values? Also, what is a way you could make this pattern accessible? Add tabindex / role? Anything else?

There is nothing inherently inaccessible about a hidden input element, even with a visible label. Perhaps there is a lightbox and all focusable elements are given `visibility:none`, perhaps there is something else going on why one is visible and the other isn't. The hidden input element is a symptom, but not the actual problem.

**Comment by marcysutton** [comment]
I did some testing around tabindex on the label, and it still wasn't enough to interact with the input. If the label had an interactive role it might work better, but it still wouldn't reflect changes to the control without managing its own `aria-checked`. That could be an exception, perhaps?

**Comment by WilcoFiers** [comment]
Yeah, I think that's an exception this would need. Any thoughts on how to deal with potential false positives? There's definitely a risk of those with this idea.

**Comment by marcysutton** [comment]
Are there any specific cases you're thinking of? We could tag the rule as experimental and play around with it in aXe Coconut.

**Comment by cyns** [comment]
I'm working on this one

**Comment by dylanb** [comment]
I am concerned about false positives for this rule

**Comment by marcysutton** [comment]
@dylanb anything specific? It would be helpful to discuss those cases now.

**Comment by dylanb** [comment]
Specifically - it is possible to use a label unsemantically and still have a completely accessible page/application. So I see the value of a best practice that flags these.

If the input is display:none, there could still be some other keyboard accessible mechanism to trigger it. Perhaps implement a separate review item rule?

**Comment by straker** [comment]
After discussion we've decided that we don't know how to solve this without generating false positives.
